### PR TITLE
add biharmonic example

### PR DESCRIPTION
**Pull request by weinmueller**
I added the biharmonic example (for a single patch) with the new expressions
and with two possible boundary conditions:
(use flag --neumann)
u = g_1
\partial_n u = g_2

and (default)
u = g_1
\Delta u = g_3

# Please consider the following checklist before issuing a pull
request:
- [X] Have you added an explanation of what your changes do and why
  you'd like us to include them?
- [X] Have you documented any new codes using Doxygen comments?
- [X] Have you written new tests or examples for your changes?
-----

### CODE REVIEW

**Comment by weinmueller** [comment]
Do we need to change the Doxygen as well?

**Comment by filiatra** [comment]
> Do we need to change the Doxygen as well?
In principle just add one line to include this source code, as with the other examples..

**Comment by weinmueller** [comment]
I think I did it. Or do you mean sth different?

File: doc/Examples.dox [hunk 1]
```diff
@@ -16,6 +16,8 @@ In the gismo/examples sub-directory we find:
 
 - \subpage biharmonic_example
 
+- \subpage biharmonic2_example
```
Comment (stefantakacs) [inline doc/Examples.dox:19]: Please add a new file `doc/biharmonic2_example.dox` based on `doc/biharmonic2_example.dox`.

You are welcome also to provide more, like in `doc/multiGrid_example.dox`

File: examples/biharmonic2_example.cpp [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+    @brief A Biharmonic example for a single patch.
+
+    This file is part of the G+Smo library.
+
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+    Author(s): J. Sogn
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: + P. Weinmüller or instead

File: examples/biharmonic2_example.cpp [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    gsCmdLine cmd("Example for solving the biharmonic problem (single patch only).");
+    cmd.addInt( "p", "discreteDegree","Discrete polynomial degree", discreteDegree );
+    cmd.addInt( "r", "discreteRegularity", "Discrete regularity",  discreteRegularity );
+    cmd.addInt( "l", "refinementLoop", "Number of refinement steps",  numRefine );
+    cmd.addString( "f", "file", "Input geometry file (with .xml)", fn );
+
+    cmd.addSwitch("last", "Solve solution for the last level of h-refinement", last);
+    cmd.addSwitch("plot", "Create a ParaView visualization file with the solution", plot);
+
+    cmd.addSwitch("neumann", "Compute problem with Neumann condition (default Laplace condition)", neumann);
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: Could we rename that?

First biharmonic:  u and \partial u/\partial n
Second biharmonic: u and Delta u

File: examples/biharmonic2_example.cpp [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    index_t discreteDegree = 3;
+    index_t discreteRegularity = 2;
+    bool last = false;
+    bool neumann = false;
+
+    std::string fn;
+
+    gsCmdLine cmd("Example for solving the biharmonic problem (single patch only).");
+    cmd.addInt( "p", "discreteDegree","Discrete polynomial degree", discreteDegree );
+    cmd.addInt( "r", "discreteRegularity", "Discrete regularity",  discreteRegularity );
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: In other files, we tend to have "degree" and "smoothness" (this both applies to the argument and the variable name)

File: examples/biharmonic2_example.cpp [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    {
+        bnd = basis.basis(it->ps.patch).boundaryOffset(it->ps.side(),1);
+        mapper.markBoundary(it->ps.patch, bnd, 0);
+    }
+    mapper.finalize();
+}
+
+void setDirichletNeumannValuesL2Projection(gsMultiPatch<> & mp, gsMultiBasis<> & basis, gsBoundaryConditions<> & bc, const expr::gsFeSpace<real_t> & u)
+{
+    gsDofMapper mapper = u.mapper();
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: We do not need a copy. Maybe make a const reference.

File: examples/biharmonic2_example.cpp [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    cmd.addSwitch("last", "Solve solution for the last level of h-refinement", last);
+    cmd.addSwitch("plot", "Create a ParaView visualization file with the solution", plot);
+
+    cmd.addSwitch("neumann", "Compute problem with Neumann condition (default Laplace condition)", neumann);
+    try { cmd.getValues(argc,argv); } catch (int rv) { return rv; }
+
+    gsMultiPatch<> mp;
+    gsBoundaryConditions<> bc;
+    gsFunctionExpr<> f, ms;
+    gsOptionList optionList;
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: unused

File: examples/biharmonic2_example.cpp [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+    //! [Read geometry]
+    std::string string_geo;
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: remove

File: examples/biharmonic2_example.cpp [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+    if (fn.empty())
+        mp = gsMultiPatch<>( *gsNurbsCreator<>::BSplineFatQuarterAnnulus() );
+    else
+    {
+        string_geo = fn;
+        gsInfo << "Filedata: " << string_geo << "\n";
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: just write fn instead of string_geo

File: examples/biharmonic2_example.cpp [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+        gsReadFile<>(string_geo, mp);
+    }
+    mp.clearTopology();
+    mp.computeTopology();
+
+    if (mp.nPatches() != 1)
+    {
+        gsInfo << "The geometry has more than one patch. Run the code with a single Patch!\n";
+        return 0;
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: return EXIT_FAILURE;

File: examples/biharmonic2_example.cpp [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+
+    cmd.addSwitch("last", "Solve solution for the last level of h-refinement", last);
+    cmd.addSwitch("plot", "Create a ParaView visualization file with the solution", plot);
+
+    cmd.addSwitch("neumann", "Compute problem with Neumann condition (default Laplace condition)", neumann);
+    try { cmd.getValues(argc,argv); } catch (int rv) { return rv; }
+
+    gsMultiPatch<> mp;
+    gsBoundaryConditions<> bc;
+    gsFunctionExpr<> f, ms;
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: Remove      gsFunctionExpr<> f, ms;

File: examples/biharmonic2_example.cpp [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+    mp.clearTopology();
+    mp.computeTopology();
+
+    if (mp.nPatches() != 1)
+    {
+        gsInfo << "The geometry has more than one patch. Run the code with a single Patch!\n";
+        return 0;
+    }
+
+    gsFunctionExpr<>source("256*pi*pi*pi*pi*(4*cos(4*pi*x)*cos(4*pi*y) - cos(4*pi*x) - cos(4*pi*y))",2);
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: source => f and remove next line

File: examples/biharmonic2_example.cpp [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+    f.swap(source);
+    gsInfo << "Source function " << f << "\n";
+
+    gsFunctionExpr<> solution("(cos(4*pi*x) - 1) * (cos(4*pi*y) - 1)",2);
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: solution => ms and remove next line

File: examples/biharmonic2_example.cpp [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+    gsMultiBasis<> basis(mp, false);//true: poly-splines (not NURBS)
+
+    // Elevate and p-refine the basis to order p + numElevate
+    // where p is the highest degree in the bases
+    basis.setDegree( discreteDegree); // preserve smoothness
+
+    // h-refine each basis
+    if (last)
+    {
+        for (int r =0; r < numRefine; ++r)
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: index_t

File: examples/biharmonic2_example.cpp [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+    //! [Solver loop]
+    gsSparseSolver<real_t>::SimplicialLDLT solver;
+
+    gsVector<real_t> l2err(numRefine+1), h1err(numRefine+1), h2err(numRefine+1),
+            dofs(numRefine+1), meshsize(numRefine+1);
+    gsInfo<< "(dot1=assembled, dot2=solved, dot3=got_error)\n"
+             "\nDoFs: ";
+    double setup_time(0), ma_time(0), slv_time(0), err_time(0);
+    gsStopwatch timer;
+    for (int r=0; r<=numRefine; ++r)
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: index_t

File: examples/biharmonic2_example.cpp [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+
+    std::string fn;
+
+    gsCmdLine cmd("Example for solving the biharmonic problem (single patch only).");
+    cmd.addInt( "p", "discreteDegree","Discrete polynomial degree", discreteDegree );
+    cmd.addInt( "r", "discreteRegularity", "Discrete regularity",  discreteRegularity );
+    cmd.addInt( "l", "refinementLoop", "Number of refinement steps",  numRefine );
+    cmd.addString( "f", "file", "Input geometry file (with .xml)", fn );
+
+    cmd.addSwitch("last", "Solve solution for the last level of h-refinement", last);
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: Solve problem only on the last level of h-refinement

File: examples/biharmonic2_example.cpp [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+        gsInfo<<"Plotting in Paraview...\n";
+        ev.options().setSwitch("plot.elements", true);
+        ev.writeParaview( u_sol   , G, "solution");
+        gsInfo <<"Saved with solution.pvd \n";
+    }
+    else
+        gsInfo << "Done. No output created, re-run with --plot to get a ParaView "
+                  "file containing the solution.\n";
+
+    return  0;
```
Comment (stefantakacs) [inline examples/biharmonic2_example.cpp]: return EXIT_SUCCESS;

**Review by stefantakacs — COMMENTED** [review]
Looks ok, I will come back to it after some local testing.

File: doc/Examples.dox (hunk 1 above)
Comment (weinmueller) [reply doc/Examples.dox:19]: Do you know how I can see the result of the dox?

Comment (stefantakacs) [reply doc/Examples.dox:19]: Should work with calling `make doc` in the build directory.

**Comment by stefantakacs** [comment]
When I run `bin/biharmonic2_example`, I get the error msg:

terminate called after throwing an instance of 'std::logic_error'
  what():  GISMO_ASSERT `m_sd->mapper.isFinalized()` The provided dof-mapper is not finalized.
/home/darwin/stefant/gismo/src/gsAssembler/gsExpressions.h, line 940 (setupMapper)
Aborted (core dumped)

**Comment by weinmueller** [comment]
> When I run `bin/biharmonic2_example`, I get the error msg:
> 
> terminate called after throwing an instance of 'std::logic_error' what(): GISMO_ASSERT `m_sd->mapper.isFinalized()` The provided dof-mapper is not finalized. /home/darwin/stefant/gismo/src/gsAssembler/gsExpressions.h, line 940 (setupMapper) Aborted (core dumped)

Done. I thought we already had this fixed.

File: doc/Examples.dox (hunk 1 above)
Comment (weinmueller) [reply doc/Examples.dox:19]: I updated the Doc.

**Comment by weinmueller** [comment]
We moved this PR to the PR #519

**Pull request closed without merging**